### PR TITLE
Feature/27/show the history of the calculator

### PR DESCRIPTION
**Pull request by yaaHosein**
Let’s work together!

### CODE REVIEW

**Comment by SMH110** [comment]
what is the reason of creating this file
image

do you want to confuse your self with it?

**Comment by yaaHosein** [comment]
> what is the reason of creating this file
> image
> 
> do you want to confuse your self with it?

No I don’t 
I removed it before seeing your comment 
I used to change any file’s name by remaining it
But maybe the git command rules are different 
When I rename a file the old one remains 
Thanks

**Comment by SMH110** [comment]
> > what is the reason of creating this file
> > image
> > do you want to confuse your self with it?
> 
> No I don’t
> I removed it before seeing your comment
> I used to change any file’s name by remaining it
> But maybe the git command rules are different
> When I rename a file the old one remains
> Thanks

This file still exist in your repo, you need to delete and commit your changes and push them

File: DevelopedCalculator.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class DevelopedCalculator extends Calculator {
+    render_() {
```
Comment (SMH110) [inline DevelopedCalculator.js]: The name of this method must bee `render` otherwise when you are doing this
`developedCalculator.render()`
the parent class render is been invoked not `DevelopedCalculator` render method.

you need to have a single interface but in side the new render method call the parent render method

**Comment by SMH110** [comment]
What you have done so far is very important.
You defined a render method inside the new class, this method will be called by the user of your code.
Which is that moment only you. But we have now a single interface (both classes have render method)
and inside the new render method, you called the parent render method 

This will give us, a full calculator ready to be used, we didn't have to do anything 
and now, our job is to add a new feature to the previous calculator.

What you're learning now is very important, this called design pattern and inheritance 


now in the new render method you need to create a new div element 
set the text of div element to say "History: " 

then prepent the new div element to the host element 


```js

const historyDiv = .... 
// add text to historyDiv
this.hostElem.prepend(historyDiv)
```
if you're wondering where from this `hostElem` came?
this is defined as property on the parent class 
this hostElem hold a reference of the HTML element which is top parent of the calculator 

and `prepend` method allows you add  a HTML element to the beginning of the `hostElem` 
while `append` does the opposite

File: DevelopedCalculator.js [hunk 2]
```diff
@@ -1,6 +1,9 @@
 class DevelopedCalculator extends Calculator {
     render() {
         super.render()
+        const history = document.createElement('div');
+        const historyDiv_ = document.createTextNode("History:");
```
Comment (SMH110) [inline DevelopedCalculator.js]: `historyDiv_` can be named (historyContent)

File: DevelopedCalculator.js [hunk 3]
```diff
@@ -1,6 +1,9 @@
...
+        const historyDiv = this.hostElm.prepend(historyDiv)
```
Comment (SMH110) [inline DevelopedCalculator.js]: how to add a text content to a HTM element?

Also when you prepent the historyDiv to the `hostElm` you don't need to assign that to a variable

File: DevelopedCalculator.js [hunk 4]
```diff
@@ -1,9 +1,11 @@
...
     render() {
         super.render()
-        const history = document.createElement('div');
-        const historyDiv_ = document.createTextNode("History:");
-        const historyDiv = this.hostElm.prepend(historyDiv)
+        const historyDiv = document.createElement('div');
+        const historyDivText = document.createTextNode("History:");
+        historyDiv_.appendChild(historyDivText);
+        this.hostElm.prepend(historyDiv);
+        historyDiv.appendChild(historyDiv_);
```
Comment (SMH110) [inline DevelopedCalculator.js]: this how you should do it:

```js
// first create a div 
const historyDiv = document.createElement("div");
// add the text to the div
historyDiv.appendChild(document.createTextNode("History:"));
// prepend the historyDiv in the host element 

this.hostElm.prepend(historyDiv);

```
does this make sense?

Comment (yaaHosein) [reply DevelopedCalculator.js]: Yes it is because it’s your code 
But why It is not looks like what you taught  me in the Calculator class 

why you did not add this code:
const historyDivText = document.createTextNode("History:");

**Comment by yaaHosein** [comment]
please I want to research about  adding text to an element :
is this site okay:
https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent

?

**Comment by yaaHosein** [comment]
Please I have tow questions also I am not urgent for their answers:
1- why we are created new file for this class 
why we did not continue working in the Calculator logic file
there you added new class (Calculator) 

====
2- why you did not added this current feature in class Calculator instead creating new class (Developed.. )
?

**Comment by SMH110** [comment]
> please I want to research about adding text to an element :
> is this site okay:
> https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent
> 
> ?

yes

**Comment by yaaHosein** [comment]
> please I want to research about adding text to an element :
> is this site okay:
> https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent
> 
> ?

Thanks

**Comment by SMH110** [comment]
is the History div showing?

**Comment by yaaHosein** [comment]
> is the History div showing?

Sorry I forgot to check 
I will check it when I will be near the laptop

**Comment by SMH110** [comment]
> Please I have tow questions also I am not urgent for their answers:
> 1- why we are created new file for this class
> why we did not continue working in the Calculator logic file
> there you added new class (Calculator)
> 
> ====
> 2- why you did not added this current feature in class Calculator instead creating new class (Developed.. )
> ?

1. You should always split your code into modules, each file should have only one class or few lines of code related to each other 
this helps in maintaining the code.
The other main reason I asked you put this class in a new file, because last time you were keep changing lines which they don't have any reason to change, so I wanted to prevent you of doing this.
I didn't want you to go and change the Calculator class.

2. This is answered above

**Comment by yaaHosein** [comment]
> > Please I have tow questions also I am not urgent for their answers:
> > 1- why we are created new file for this class
> > why we did not continue working in the Calculator logic file
> > there you added new class (Calculator)
> > ====
> > 2- why you did not added this current feature in class Calculator instead creating new class (Developed.. )
> > ?
> 
> 1. You should always split your code into modules, each file should have only one class or few lines of code related to each other
>    this helps in maintaining the code.
>    The other main reason I asked you put this class in a new file, because last time you were keep changing lines which they don't have any reason to change, so I wanted to prevent you of doing this.
>    I didn't want you to go and change the Calculator class.
> 2. This is answered above

Ok thanks a lot 
It makes sense

**Comment by yaaHosein** [comment]
Thanks

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by yaaHosein** [comment]
> is the History div showing?

Unfortunately I can’t see it

**Comment by yaaHosein** [comment]
po

**Comment by SMH110** [comment]
Thank you for investigating the issue.
it's saying can not ready property "prepend" of undefined and it's giving you the line number where this error occurred (line 9)
which is this one 
`this.hostElm.prepend(historyDiv);`

`hostElm` is undefined, hence it does not have "prepend" property.

I had a look at the Calculator class and found the Calculator class has this code
`this.hostElem = document.querySelector(selector);`
here where you define the "hostElem" 
and if you compare "hostElem" with `hostElm` you will find the typo 

I would expect you to debug these type of issues same as I did now.

**Comment by SMH110** [comment]
After fixing the issue which stopping you from seeing the Hisotry `div`
I want to not only show "History:" word like this.
I want you to show the History + the current result of the calculator.

the current result is available on the parent class

**Comment by yaaHosein** [comment]
> Thank you for investigating the issue.
> it's saying can not ready property "prepend" of undefined and it's giving you the line number where this error occurred (line 9)
> which is this one
> `this.hostElm.prepend(historyDiv);`
> 
> `hostElm` is undefined, hence it does not have "prepend" property.
> 
> I had a look at the Calculator class and found the Calculator class has this code
> `this.hostElem = document.querySelector(selector);`
> here where you define the "hostElem"
> and if you compare "hostElem" with `hostElm` you will find the typo
> 
> I would expect you to debug these type of issues same as I did now.

thanks a lot
I tried to check hostElm but in the html file only

**Comment by yaaHosein** [comment]
> After fixing the issue which stopping you from seeing the Hisotry `div`
> I want to not only show "History:" word like this.
> I want you to show the History + the current result of the calculator.
> 
> the current result is available on the parent class

History: is above the current result now
do you mean do you want to see both of them in one line ?

**Comment by SMH110** [comment]
> History: is above the current result now
> do you mean do you want to see both of them in one line ?
> 

When you open the app, you see the result div saying "0"
I want you to make the History section says "History: 0" , but I don't want you hardcoded "0" , instead I want you to use the currentResult property for that

**Comment by SMH110** [comment]
> Please I have tow questions also I am not urgent for their answers:
> 1- why we are created new file for this class
> why we did not continue working in the Calculator logic file
> there you added new class (Calculator)
> 
> ====
> 2- why you did not added this current feature in class Calculator instead creating new class (Developed.. )
> ?


sorry, I noticed that your question 2 is different 
The main reason we are trying to show the history in another class is to make it possible using the calculator without the history.

Let's say you're building these calculators for many users, and you want to give your users the options to choose the features they want.
some users might want to use calculator + history and some might want to use the calculator app only.

the other reason, so that you learn about inheritance and some design pattern

**Comment by yaaHosein** [comment]
> > History: is above the current result now
> > do you mean do you want to see both of them in one line ?
> > 
> 
> When you open the app, you see the result div saying "0"
> I want you to make the History section says "History: 0" , but I don't want you hardcoded "0" , instead I want you to use the currentResult property for that


I do not know what happened to this property 
I am adding this.currentResult to the History: 
But it shows 
History:0

I did not add “0”

I know the current result is 0 but I am using this 

Despite I tried another way by adding this.update.. method 

But still now this property is not showing the new current results

**Comment by yaaHosein** [comment]
> > Please I have tow questions also I am not urgent for their answers:
> > 1- why we are created new file for this class
> > why we did not continue working in the Calculator logic file
> > there you added new class (Calculator)
> > ====
> > 2- why you did not added this current feature in class Calculator instead creating new class (Developed.. )
> > ?
> 
> sorry, I noticed that your question 2 is different
> The main reason we are trying to show the history in another class is to make it possible using the calculator without the history.
> 
> Let's say you're building these calculators for many users, and you want to give your users the options to choose the features they want.
> some users might want to use calculator + history and some might want to use the calculator app only.
> 
> the other reason, so that you learn about inheritance and some design pattern

Thanks a lot

**Comment by SMH110** [comment]
> I do not know what happened to this property
> I am adding this.currentResult to the History:
> But it shows
> History:0
> 
> I did not add “0”
> 
> I know the current result is 0 but I am using this
> 
> Despite I tried another way by adding this.update.. method
> 
> But still now this property is not showing the new current results

can i see your code?

**Comment by SMH110** [comment]
This what I am seeing now.
image

which is correct.

I think your question was:
I don't know why it's showing  History: 0 **while the result div showing 07**
image

this is fine for now, we will fix this later.
to explain what happening 
when initially you call the `render` method, the calculator ui gets build and injected to the page by this line 
`super.render()`
then you add the history div element to the host element of the calculator.
initally the `currentResult` which's coming from the parent class is set to `0`

then when you click on button 7, the currentResult is set to `07`, but did you update the history div to get the recent value of the currentResult?

no, and this what will be doing next

**Comment by yaaHosein** [comment]
Thanks for the detailed answer 
In the beginning I thought this. property will update the result div

But I noticed it’s not 
So I added update result method 
Also it did not do what I was looking for

**Comment by SMH110** [comment]
Imagine, there's a queue of students.
students are walking through that queue to receive their exam result.

Your job is to identify the top student and before they receive their result you need to pick a gift and hand over the gift with the result.

in this case, you intercept the process of result handover to decorate the result with a gift.

This what we need to do in other to implement the history feature.

when a user click on a button, we intercept the click handler, do something and then call the click handler.

Remember that we have `onButtonNumberClick` on the Calculator class.
this method is the main method that updates the `currentResult` 

we need to intercept this method and do some extra stuff to allow of showing the history 

first, you need to add to the new class an array 
this array will keep track of the currentResult and show it changes 

then create `onButtonNumberClick` on the new class 
it's very important that the name of the method matches with the method on the parent class 

this method (onButtonNumberClick) will receive the event object which need to pass on the method on the parent class
`super.onButtonNumberClick(event)`

then, add the current result to the history array 

`this.history.push(this.currentResult)`


when a user clicks on button 1
JS will start from the child class searching for `onButtonNumberClick` method to execute it; because of this line
`  buttonElement.addEventListener("click", this.onButtonNumberClick);`

`this` keyword here, refers to the instance of `DevelopedCalculator`. the engine will find the method defined on the `DevelopedCalculator` so it will pass it the event object.

Before adding this method, the calculator is still working fine, because the engine will try to find the method on `DevelopedCalculator` and it's not finding it, in this case it will start searching on the parent class and so on

**Comment by SMH110** [comment]
image

**Comment by yaaHosein** [comment]
nb

**Comment by yaaHosein** [comment]
I do not know how it is deleted 
can I add these two files 
?

**Comment by yaaHosein** [comment]
I did this change in the CalculatorLOgic file:
14l

I have deleted the second one

**Comment by SMH110** [comment]
> I did this change in the CalculatorLOgic file:
> 14l
> 
> I have deleted the second one

Thanks for asking before committing changes to `CalculaorLogic.js` 
Please remove line 191 and line 192 then commit your changes. 
just so we have 1 calculator in the UI

**Comment by SMH110** [comment]
https://www.youtube.com/watch?v=3Tn58KQvWtU&ab_channel=DevWorld

**Comment by yaaHosein** [comment]
> Imagine, there's a queue of students.
> students are walking through that queue to receive their exam result.
> 
> Your job is to identify the top student and before they receive their result you need to pick a gift and hand over the gift with the result.
> 
> in this case, you intercept the process of result handover to decorate the result with a gift.
> 
> This what we need to do in other to implement the history feature.
> 
> when a user click on a button, we intercept the click handler, do something and then call the click handler.
> 
> Remember that we have `onButtonNumberClick` on the Calculator class.
> this method is the main method that updates the `currentResult`
> 
> we need to intercept this method and do some extra stuff to allow of showing the history
> 
> first, you need to add to the new class an array
> this array will keep track of the currentResult and show it changes
> 
> then create `onButtonNumberClick` on the new class
> it's very important that the name of the method matches with the method on the parent class
> 
> this method (onButtonNumberClick) will receive the event object which need to pass on the method on the parent class
> `super.onButtonNumberClick(event)`
> 
> then, add the current result to the history array
> 
> `this.history.push(this.currentResult)`
> 
> when a user clicks on button 1
> JS will start from the child class searching for `onButtonNumberClick` method to execute it; because of this line
> ` buttonElement.addEventListener("click", this.onButtonNumberClick);`
> 
> `this` keyword here, refers to the instance of `DevelopedCalculator`. the engine will find the method defined on the `DevelopedCalculator` so it will pass it the event object.
> 
> Before adding this method, the calculator is still working fine, because the engine will try to find the method on `DevelopedCalculator` and it's not finding it, in this case it will start searching on the parent class and so on

Regarding to 
Keep track of 

Do you mean to do something like what you mentioned long time ago 
In the following email ((12 Dec 2020) when you were explaining keep track of:



I have added a new ticket for you 
https://github.com/yaaHosein/Calculator1/issues/13
After reading & understanding the requirements in the link above; continue reading this email

In order to implement this new feature you need to do few things 

First, you need to remove that code which clear the result div when operator button clicked. 

I can see these two lines in the addingButton and subractingButton clearing the result div 
Image

Second, you need to keep track of the last button was clicked every time user clicks on a button 
for example, if a user clicks on 1 button you need to store this in a variable and use this variable to hold the type of the button was clicked 
every time the user clicks on a number button (1,2,3,4,5,...etc) you check if the variable holding the last button clicked is an operator button 
and if this equal to true then clear the result div. and this ensures that user last number not added to the first number


What need to be done?

create a variable that store the value of the last button clicked and make sure the variable is created somewhere that can be assessed in every function.
update all buttons event handlers to update the variable created in the step 1, for example when user clicks on button 2 the variable value should have value"2" or "NUMBER" and when a + button clicked the variable should have value of "+" 
update all numbers event handlers to start checking if the value of the button created at step 1 is equal to an operator type for example "+" or "-" or "*" ...etc
if this true then clear the result div 
Make sure to remove the code which clearing the result div inside the operators event handlers

**Comment by SMH110** [comment]
> > Imagine, there's a queue of students.
> > students are walking through that queue to receive their exam result.
> > Your job is to identify the top student and before they receive their result you need to pick a gift and hand over the gift with the result.
> > in this case, you intercept the process of result handover to decorate the result with a gift.
> > This what we need to do in other to implement the history feature.
> > when a user click on a button, we intercept the click handler, do something and then call the click handler.
> > Remember that we have `onButtonNumberClick` on the Calculator class.
> > this method is the main method that updates the `currentResult`
> > we need to intercept this method and do some extra stuff to allow of showing the history
> > first, you need to add to the new class an array
> > this array will keep track of the currentResult and show it changes
> > then create `onButtonNumberClick` on the new class
> > it's very important that the name of the method matches with the method on the parent class
> > this method (onButtonNumberClick) will receive the event object which need to pass on the method on the parent class
> > `super.onButtonNumberClick(event)`
> > then, add the current result to the history array
> > `this.history.push(this.currentResult)`
> > when a user clicks on button 1
> > JS will start from the child class searching for `onButtonNumberClick` method to execute it; because of this line
> > ` buttonElement.addEventListener("click", this.onButtonNumberClick);`
> > `this` keyword here, refers to the instance of `DevelopedCalculator`. the engine will find the method defined on the `DevelopedCalculator` so it will pass it the event object.
> > Before adding this method, the calculator is still working fine, because the engine will try to find the method on `DevelopedCalculator` and it's not finding it, in this case it will start searching on the parent class and so on
> 
> Regarding to
> Keep track of
> 
> Do you mean to do something like what you mentioned long time ago
> In the following email ((12 Dec 2020) when you were explaining keep track of:
> 
> I have added a new ticket for you
> #13
> After reading & understanding the requirements in the link above; continue reading this email
> 
> In order to implement this new feature you need to do few things
> 
> First, you need to remove that code which clear the result div when operator button clicked.
> 
> I can see these two lines in the addingButton and subractingButton clearing the result div
> Image
> 
> Second, you need to keep track of the last button was clicked every time user clicks on a button
> for example, if a user clicks on 1 button you need to store this in a variable and use this variable to hold the type of the button was clicked
> every time the user clicks on a number button (1,2,3,4,5,...etc) you check if the variable holding the last button clicked is an operator button
> and if this equal to true then clear the result div. and this ensures that user last number not added to the first number
> 
> What need to be done?
> 
> create a variable that store the value of the last button clicked and make sure the variable is created somewhere that can be assessed in every function.
> update all buttons event handlers to update the variable created in the step 1, for example when user clicks on button 2 the variable value should have value"2" or "NUMBER" and when a + button clicked the variable should have value of "+"
> update all numbers event handlers to start checking if the value of the button created at step 1 is equal to an operator type for example "+" or "-" or "*" ...etc
> if this true then clear the result div
> Make sure to remove the code which clearing the result div inside the operators event handlers

To keep track means to watch something or series of values 

every time a button is clicked we need to keep a record of that so we can show it in the history.

for example when button 1 is clicked 
the history should show

History: 1 

when the button 2 is clicked, the history should show

History: 1, 12 

and when button 2 clicked again
the history should should show:

History: 1, 12, 122

in this case we need to watch the buttons (keep track) and we want to watch the values through out the time
in other words, we are not watching only 1 value, we are watching a series of values.
we are not interested in the most recent value of the `currentResult`, but we are interested in the changes happened to the `currentResult` 

so, we need to use an array for that.
the array can store the values of the `currentResult` every time the currentResult change 

`[1, 12, 122] `

**Comment by yaaHosein** [comment]
> > > Imagine, there's a queue of students.
> > > students are walking through that queue to receive their exam result.
> > > Your job is to identify the top student and before they receive their result you need to pick a gift and hand over the gift with the result.
> > > in this case, you intercept the process of result handover to decorate the result with a gift.
> > > This what we need to do in other to implement the history feature.
> > > when a user click on a button, we intercept the click handler, do something and then call the click handler.
> > > Remember that we have `onButtonNumberClick` on the Calculator class.
> > > this method is the main method that updates the `currentResult`
> > > we need to intercept this method and do some extra stuff to allow of showing the history
> > > first, you need to add to the new class an array
> > > this array will keep track of the currentResult and show it changes
> > > then create `onButtonNumberClick` on the new class
> > > it's very important that the name of the method matches with the method on the parent class
> > > this method (onButtonNumberClick) will receive the event object which need to pass on the method on the parent class
> > > `super.onButtonNumberClick(event)`
> > > then, add the current result to the history array
> > > `this.history.push(this.currentResult)`
> > > when a user clicks on button 1
> > > JS will start from the child class searching for `onButtonNumberClick` method to execute it; because of this line
> > > ` buttonElement.addEventListener("click", this.onButtonNumberClick);`
> > > `this` keyword here, refers to the instance of `DevelopedCalculator`. the engine will find the method defined on the `DevelopedCalculator` so it will pass it the event object.
> > > Before adding this method, the calculator is still working fine, because the engine will try to find the method on `DevelopedCalculator` and it's not finding it, in this case it will start searching on the parent class and so on
> > 
> > 
> > Regarding to
> > Keep track of
> > Do you mean to do something like what you mentioned long time ago
> > In the following email ((12 Dec 2020) when you were explaining keep track of:
> > I have added a new ticket for you
> > #13
> > After reading & understanding the requirements in the link above; continue reading this email
> > In order to implement this new feature you need to do few things
> > First, you need to remove that code which clear the result div when operator button clicked.
> > I can see these two lines in the addingButton and subractingButton clearing the result div
> > Image
> > Second, you need to keep track of the last button was clicked every time user clicks on a button
> > for example, if a user clicks on 1 button you need to store this in a variable and use this variable to hold the type of the button was clicked
> > every time the user clicks on a number button (1,2,3,4,5,...etc) you check if the variable holding the last button clicked is an operator button
> > and if this equal to true then clear the result div. and this ensures that user last number not added to the first number
> > What need to be done?
> > create a variable that store the value of the last button clicked and make sure the variable is created somewhere that can be assessed in every function.
> > update all buttons event handlers to update the variable created in the step 1, for example when user clicks on button 2 the variable value should have value"2" or "NUMBER" and when a + button clicked the variable should have value of "+"
> > update all numbers event handlers to start checking if the value of the button created at step 1 is equal to an operator type for example "+" or "-" or "*" ...etc
> > if this true then clear the result div
> > Make sure to remove the code which clearing the result div inside the operators event handlers
> 
> To keep track means to watch something or series of values
> 
> every time a button is clicked we need to keep a record of that so we can show it in the history.
> 
> for example when button 1 is clicked
> the history should show
> 
> History: 1
> 
> when the button 2 is clicked, the history should show
> 
> History: 1, 12
> 
> and when button 2 clicked again
> the history should should show:
> 
> History: 1, 12, 122
> 
> in this case we need to watch the buttons (keep track) and we want to watch the values through out the time
> in other words, we are not watching only 1 value, we are watching a series of values.
> we are not interested in the most recent value of the `currentResult`, but we are interested in the changes happened to the `currentResult`
> 
> so, we need to use an array for that.
> the array can store the values of the `currentResult` every time the currentResult change
> 
> `[1, 12, 122]`

Okay thank you so much for your nice detailed explanation

**Comment by SMH110** [comment]
Any update on this?

**Comment by yaaHosein** [comment]
Sorry for the late reply 
I think we can use 
ForEach method 

But how ?!

هون حطنا الجمّال

**Comment by SMH110** [comment]
> Sorry for the late reply
> I think we can use
> ForEach method
> 
> But how ?!
> 
> هون حطنا الجمّال

Why do you want to use `forEach` ?
where have you stuck?

**Comment by yaaHosein** [comment]
> > Sorry for the late reply
> > I think we can use
> > ForEach method
> > But how ?!
> > هون حطنا الجمّال
> 
> Why do you want to use `forEach` ?
> where have you stuck?
1) for showing the series of events 
2)
For example in changing the value of history 
I mean changing the 0 to current result

**Comment by SMH110** [comment]
> > > Sorry for the late reply
> > > I think we can use
> > > ForEach method
> > > But how ?!
> > > هون حطنا الجمّال
> > 
> > 
> > Why do you want to use `forEach` ?
> > where have you stuck?
> 
> 1. for showing the series of events
> 
> For example in changing the value of history
> I mean changing the 0 to current result

`forEach` is used to loop over any array 

First you need the array to store the current results changes in it?
Have you created the array?

**Comment by yaaHosein** [comment]
> > > > Sorry for the late reply
> > > > I think we can use
> > > > ForEach method
> > > > But how ?!
> > > > هون حطنا الجمّال
> > > 
> > > 
> > > Why do you want to use `forEach` ?
> > > where have you stuck?
> > 
> > 
> > 
> > 1. for showing the series of events
> > 
> > For example in changing the value of history
> > I mean changing the 0 to current result
> 
> `forEach` is used to loop over any array
> 
> First you need the array to store the current results changes in it?
> Have you created the array?

Yes since some days ago 
Basically I told you maybe I have to use for Each function for filling this empty array []

**Comment by SMH110** [comment]
> > > > > Sorry for the late reply
> > > > > I think we can use
> > > > > ForEach method
> > > > > But how ?!
> > > > > هون حطنا الجمّال
> > > > 
> > > > 
> > > > Why do you want to use `forEach` ?
> > > > where have you stuck?
> > > 
> > > 
> > > 
> > > 1. for showing the series of events
> > > 
> > > For example in changing the value of history
> > > I mean changing the 0 to current result
> > 
> > 
> > `forEach` is used to loop over any array
> > First you need the array to store the current results changes in it?
> > Have you created the array?
> 
> Yes since some days ago
> Basically I told you maybe I have to use for Each function for filling this empty array []

Can you give the line number when you created the array, because I can't find anything in your repo

File: DevelopedCalculator.js [hunk 5]
```diff
@@ -9,9 +10,13 @@ class DevelopedCalculator extends Calculator {
         this.hostElem.prepend(historyDiv);
 
 
-        this.updateResultElement();
+        const currentResultUpdate = []
+    }
 
+    onButtonNumberClick_(event) {
```
Comment (SMH110) [inline DevelopedCalculator.js]: Can you tell me why the method name ends with _ ?

File: DevelopedCalculator.js [hunk 6]
```diff
@@ -9,9 +10,13 @@ class DevelopedCalculator extends Calculator {
         this.hostElem.prepend(historyDiv);
 
 
-        this.updateResultElement();
+        const currentResultUpdate = []
```
Comment (SMH110) [inline DevelopedCalculator.js]: is this the array that meant to hold the values of the `currentResult`

File: DevelopedCalculator.js [hunk 7]
```diff
@@ -9,9 +10,13 @@ class DevelopedCalculator extends Calculator {
...
+    }
 
+    onButtonNumberClick_(event) {
+        super.history.push(this.currentResult)
```
Comment (SMH110) [inline DevelopedCalculator.js]: Where `super.history` is defined?

File: DevelopedCalculator.js (hunk 6 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: yes

Comment (SMH110) [reply DevelopedCalculator.js]: will you need to access this array in `onButtonNumberClick_` when a button number clicked?

File: DevelopedCalculator.js (hunk 7 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: I was not know super needs defining 

because you did not told me to define in when I added : line 4
  super.render()

Comment (SMH110) [reply DevelopedCalculator.js]: `super` refers to the parent class 
is `history` defined in the super class ?

File: DevelopedCalculator.js [hunk 8]
```diff
@@ -1,6 +1,7 @@
 class DevelopedCalculator extends Calculator {
 
     render() {
+        super.history = null;
```
Comment (SMH110) [inline DevelopedCalculator.js]: why `history` is defined on the `super` not on the `DevelopedCalculator`?
you must have reason for that?

File: DevelopedCalculator.js (hunk 5 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: then create onButtonNumberClick on the new class
it's very important that the name of the method matches with the method on the parent class

Comment (SMH110) [reply DevelopedCalculator.js]: so did you follow what I said "the name of the method matches with the method on the parent class"

File: DevelopedCalculator.js (hunk 8 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: I did not understand your first question!

Comment (SMH110) [reply DevelopedCalculator.js]: `super` refers to the parent class. in this case the super/parent class is `Calculator` class.
my question why you're defining the `history` property on the parent class not `DevelopedCalculator`

I assume you have a reason for doing this?

File: DevelopedCalculator.js (hunk 7 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: I can not see any class has super name

Comment (yaaHosein) [reply DevelopedCalculator.js]: I can not see any class has super name

Comment (SMH110) [reply DevelopedCalculator.js]: `super` is keyword, refers to the parent class 

your class `extends` `Calculator` class 
`super` keyword allow you access properties and method on the parent class 

`super.render()`
you called the `render` method of `Calculator` class

File: DevelopedCalculator.js [hunk 9]
```diff
@@ -1,5 +1,5 @@
 class DevelopedCalculator extends Calculator {
-    history = null;
+    history = 0;
```
Comment (SMH110) [inline DevelopedCalculator.js]: this is array 
`history = [ ] `

Comment (SMH110) [reply DevelopedCalculator.js]: it needs to be array to hold more than on value 

everytime, a button is clicked, we need to push new value to this array 

the new value is `this.currentResult`

Comment (SMH110) [reply DevelopedCalculator.js]: does this make sense?

File: DevelopedCalculator.js (hunk 7 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: Thanks a lot for clearing

File: DevelopedCalculator.js (hunk 9 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: Yes

File: DevelopedCalculator.js [hunk 10]
```diff
@@ -1,5 +1,5 @@
 class DevelopedCalculator extends Calculator {
-    history = [];
+    history = this.currentResult;
```
Comment (SMH110) [inline DevelopedCalculator.js]: why did you change this to `this.currentResult`?

File: DevelopedCalculator.js (hunk 7 above)
Comment (SMH110) [reply DevelopedCalculator.js]: any update on this?

File: DevelopedCalculator.js (hunk 10 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: @SMH110 SMH110 4 hours ago 
it needs to be array to hold more than on value

every time, a button is clicked, we need to push new value to this array

the new value is this.currentResult

Comment (SMH110) [reply DevelopedCalculator.js]: Here you define the array, in the method which will be called when the buttons are clicked you push the current result to this array

File: DevelopedCalculator.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+        super.render()
+            // first create a div 
+        const historyDiv = document.createElement("div");
+        // add the text to the div
+        historyDiv.appendChild(document.createTextNode("History:" + this.currentResult));
+        // prepend the historyDiv in the host element 
+        this.hostElem.prepend(historyDiv);
+    }
+
+    onButtonNumberClick_(event) {
```
Comment (SMH110) [inline DevelopedCalculator.js]: any update on the name of this method?

File: DevelopedCalculator.js [hunk 12]
```diff
@@ -11,7 +11,7 @@ class DevelopedCalculator extends Calculator {
     }
 
     onButtonNumberClick_(event) {
-        this.super.history.push(this.currentResult)
+        super.this.history.push(this.currentResult)
```
Comment (SMH110) [inline DevelopedCalculator.js]: the `history` property which is an array is defined on the `DevelopedCalculator` class
how do you access a property on a class?

do you use `this` or `super` or both? 
or you just do some random thing hopping it will work?

File: DevelopedCalculator.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+        const historyDiv = document.createElement("div");
+        // add the text to the div
+        historyDiv.appendChild(document.createTextNode("History:" + this.currentResult));
+        // prepend the historyDiv in the host element 
+        this.hostElem.prepend(historyDiv);
+    }
+
+    onButtonNumberClick_(event) {
+        super.this.history.push(this.currentResult)
+        this.updateResultElement();
```
Comment (SMH110) [inline DevelopedCalculator.js]: this wont work, you don't need to call this method here

File: DevelopedCalculator.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+        super.render()
+            // first create a div 
+        const historyDiv = document.createElement("div");
+        // add the text to the div
+        historyDiv.appendChild(document.createTextNode("History:" + this.currentResult));
+        // prepend the historyDiv in the host element 
+        this.hostElem.prepend(historyDiv);
+    }
+
+    onButtonNumberClick_(event) {
```
Comment (SMH110) [inline DevelopedCalculator.js]: why the method name ends with _ ?
any update? 
why you start doing other stuff before finishing working on this?

File: DevelopedCalculator.js (hunk 12 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: then, add the current result to the history array

this.history.push(this.currentResult)

Comment (SMH110) [reply DevelopedCalculator.js]: this `this.history.push(this.currentResult)` is correct
my question why you're using `super.this` where did you find this?

Comment (yaaHosein) [reply DevelopedCalculator.js]: sorry I was thinking you told me to add :
super.history.push(this.currentResult)

and I added what i thought very long time ago

https://github.com/yaaHosein/Calculator1/pull/29#issuecomment-822770187

Comment (SMH110) [reply DevelopedCalculator.js]: Please read carefully

File: DevelopedCalculator.js (hunk 11 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: then create onButtonNumberClick on the new class
it's very important that the name of the method matches with the method on the parent class

which name I have to choose more near to what I chosen ?

File: DevelopedCalculator.js (hunk 14 above)
Comment (yaaHosein) [reply DevelopedCalculator.js]: you are asking me for updates

Comment (SMH110) [reply DevelopedCalculator.js]: yes, because you don't reply on my question, so I have to remind you
now what happen to the method name? why still ends with _ ?
when you are going to answer?

**Comment by yaaHosein** [comment]
https://github.com/yaaHosein/Calculator1/pull/29#discussion_r619661928

**Comment by yaaHosein** [comment]
sorry today the internet is very slow most of the time 
maybe more than 10 minutes i need to wait to load github site especially if it has a lot of comments and pictures 

I am reading the following comments in my email 
not in github:

yes, because you don't reply on my question, so I have to remind you
now what happen to the method name? why still ends with _ ?
when you are going to answer?

I do not know when the internet will allow to add them in gitHub

**Comment by SMH110** [comment]
> sorry today the internet is very slow most of the time
> maybe more than 10 minutes i need to wait to load github site especially if it has a lot of comments and pictures
> 
> I am reading the following comments in my email
> not in github:
> 
> yes, because you don't reply on my question, so I have to remind you
> now what happen to the method name? why still ends with _ ?
> when you are going to answer?
> 
> I do not know when the internet will allow to add them in gitHub

If you have internet problem, you can close this PR , I mean merge it and create another one for the new changes you will add later.



before you merge your PR, I want you to change `onButtonNumberClick_` to `onButtonNumberClick`
